experimental/algorithm/LAGraph_CFL_reachability.c

```diff
@@ -0,0 +1,253 @@
+//------------------------------------------------------------------------------
+// LAGraph_CFL_reachability.c: Context-Free Language Reachability Matrix-Based Algorithm
+//------------------------------------------------------------------------------
+//
+// LAGraph, (c) 2019-2024 by The LAGraph Contributors, All Rights Reserved.
+// SPDX-License-Identifier: BSD-2-Clause
+
+// Contributed by Ilhom Kombaev, Semyon Grigoriev, St. Petersburg State University.
+
+//------------------------------------------------------------------------------
+
+// Code is based on the "A matrix-based CFPQ algorithm" described in the following paper:
+//  * Rustam Azimov, Semyon Grigorev, "Context-Free Path Querying Using Linear Algebra"
+//  * URL: https://disser.spbu.ru/files/2022/disser_azimov.pdf
+
+#define ERROR_RULE(msg)                                                                  \
+    {                                                                                    \
+        LG_ASSERT_MSGF(false, GrB_INVALID_VALUE, "Rule with index %ld is invalid. " msg, \
+                       i);                                                               \
+    }
+
+#define LG_FREE_WORK                                                                     \
+    do {                                                                                 \
+        free(nnz);                                                                       \
+    } while (0)
+
+#define LG_FREE_ALL                                                                      \
+    {                                                                                    \
+        LG_FREE_WORK;                                                                    \
+                                                                                         \
+        for (size_t i = 0; i < T_size; i++) {                                            \
+            GrB_free(&T[i]);                                                             \
+        }                                                                                \
+    }
+
+#include "LG_internal.h"
+#include <LAGraphX.h>
+
+// LAGraph_CFL_reachability: Context-Free Language Reachability Matrix-Based
+// Algorithm
+//
+// Determines the set of vertex pairs (u, v) in graph (which represented by adjacency
+// matrices) such that there is a path from u to v whose edge labels form a word from
+// the language generated by the context-free grammar (which represented by `rules`).
+//
+// Terminals and non-terminals enumerated by integers starting with zero.
+// Start nonterminal is nonterminal with index 0
+//
+// Example:
+//
+// Graph:
```

Comment:
May we we replace edge 5 -[b]-> 3 with 5 -[b]-> 2 to make the example less trivial?